### PR TITLE
UI render if salt-formulas folders are unreachable (bsc#1142309)

### PR DESCRIPTION
**Pull request by rjmateus**
## What does this PR change?

If one of the salt-formulas folders doesn't exist or have the wrong permission (755) the UI doesn't render after login. This PR ensures that the UI render as no formula exists if the folder is unreachable. 
This PR changes (Organized by commit):
- Menu entries from formulas are always visible.
- Validate folder by folder if SUMA has access to it. If just one of them have problems, the other will load as usual and display the existing formulas. 
- Log information if one some of the folders are unreachable.
- Display information in UI if some the folders are unreachable.
- Refactor in UI component to be according the new standards (Thanks @lneves12 )

## GUI diff

Change UI to display a warning message if some of the salt-formula folders are unreachable.
Before:
It was a blank page.

After:
This affects 3 pages
* Salt -> Formula Catalog
Folder OK:
image

One folder with problems:
image

Two folders with problems:
image

* System -> Minion -> Formulas
OK:
image

With ERROR:
image

* System -> Group -> Formulas
OK:
image

With ERROR:
image

- [x] **DONE**

## Documentation
- No documentation needed: Just solving a bug

- [x] **DONE**

## Test coverage
- No tests: No tests for the file, since it depends on a folder in the local disk, that would be hard to mock and ensure it will work in any context.

- [x] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/8373
Tracks # 
Manager 4.0: https://github.com/SUSE/spacewalk/pull/8990
manager 3.2: https://github.com/SUSE/spacewalk/pull/8991

- [ ] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"		 
- [ ] Re-run test "java_pgsql_tests"		 
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"		 
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"		 
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java [hunk 1]
```diff
@@ -157,10 +157,10 @@ public static String getServerDataFile() {
         File standaloneDir = new File(METADATA_DIR_STANDALONE_SALT);
         File managerDir = new File(METADATA_DIR_MANAGER);
         File customDir = new File(METADATA_DIR_CUSTOM);
-        List<File> files = new LinkedList<>(
-                Arrays.asList(standaloneDir.listFiles()));
-        files.addAll(Arrays.asList(managerDir.listFiles()));
-        files.addAll(Arrays.asList(customDir.listFiles()));
+        List<File> files = new LinkedList<>();
+        Optional.ofNullable(standaloneDir.listFiles()).ifPresent(l-> files.addAll(Arrays.asList(l)));
```
Comment (lneves12) [inline java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java]: just nitpicking: to be more readable I would change the local function variable "l" to standaloneFiles.

```java
Optional.ofNullable(standaloneDir.listFiles()).ifPresent(standaloneFiles -> files.addAll(Arrays.asList(standaloneFiles)));
Optional.ofNullable(managerDir.listFiles()).ifPresent(managerFiles-> files.addAll(Arrays.asList(managerFiles)));
```

**Comment by mcalmer** [comment]
Can we just check if this optional "IS" empty and in this case display a warning message?
I do not like blank screens and you do not know why. But also @admd is right that we hide some things which should be noted somewhere for people searching for the error.

**Comment by lneves12** [comment]
I agree with you @admd but leaving the whole UI completely unavailable because someone messed up the permissions of a folder is also not nice, so imo the suggestion from Michael would be the best approach! If the optional is empty we could just log an error

**Comment by mcalmer** [comment]
@lneves12 not just logging. Also show it in the UI and if we know which directory is messed up we can even write this. This would be coooool :-)

**Comment by rjmateus** [comment]
Thanks for the review.
I really think that we should check if the directory exists/have permission, and deal with it, instead of just show a blank page (as we have right now) where users don't know what to do or what is happening. Even the error message in the logs is not very helpfull. 

The goal is changed from "one formula folder with problems, no suse manager" to "one formula folder with problems, don't show the formulas in that folder but suse manager keeps working as usual"

For this reason, I think we really should keep the validation. But as is, I'm hiding the real problem and should somehow notify the user/admin that the folder is unreachable, and which folder have the problem.  
This block of code run's every time we load the menu, and in the current implementation (without the SPA PR from Luis) is always we load a page. 

I will add the information to log. Should it be ERROR or WARNING level? I think it should be an error since is a scenario that should not happen.
Add it to the UI, every time the page loads, is not too intrusive?

@admd , @mcalmer @lneves12  What if we always add the formulas to the menu, and show the error message in the UI only in the salt formulas page? That will prevent us from showing to many time, showing only in the context that it is part of. With this, we reduce also load time of the menu, since at each time the menu load it checks the file system.

**Comment by mcalmer** [comment]
Hiding the formulas was a feature in the beginning when we did not really had formulas with forms and wanted that feature only for consulting .

But now it does not make so much sense to hide it, as we install formuals by default.

I think it makes sense to show it always and show the error message when you want to work with formulas.

**Comment by rjmateus** [comment]
Hey @admd @mcalmer and @lneves12 
Just finished the PR, can you guys have a look? My last two commits change some UI structure according to an indication from Luis. 
@lneves12  can you check if everything is as expected?
@mcalmer the UI display the folder(s) with problems, and an indication of how to solve the issue.

File: java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java [hunk 2]
```diff
@@ -149,6 +149,28 @@ public static String getServerDataFile() {
         return dataDir + SERVER_DATA_FILE;
     }
 
+    /**
+     * Return a warning message in case some folder doesn't exist or have wrong access level.
+     * @return a warning message if cannot access one folder. NULL if all folder are ok.
+     */
+    public static String getWarningMessageAccessFormulaFolders() {
+        String message = "System folder(s) from salt-formula are unreachable. " +
+                "Check if folders(s) exists and have the correct permissions (755):";
```
Comment (admd) [inline java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java]: Please move this message to StringResource_en_US.xml in branding.

Comment (rjmateus) [reply java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java]: Thank's Abid, just change it

File: java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java [hunk 3]
```diff
@@ -174,6 +197,18 @@ public static String getServerDataFile() {
         return FormulaFactory.orderFormulas(formulasList);
     }
 
+    private static List<File> getFormulasFiles(File formulasFolder) {
```
[the remark this answers is not in the document]
Comment (rjmateus) [reply java/code/src/com/redhat/rhn/domain/formula/FormulaFactory.java]: Thanks for the suggestion. Just change it

**Comment by rjmateus** [comment]
all approved. going to merge

**Pull request merged**